powerstore/resource_host.go

```diff
@@ -277,14 +413,179 @@ func (r resourceHost) serverToState(plan, state *models.Host, response gopowerst
+	// Passwords are not queryable so in order to maintain state it is taken from the plan.
+	planInitiatorMap := make(map[types.String]models.InitiatorCreateModify)
+	if plan != nil && len(plan.Initiators) != 0 {
+		for i := 0; i < len(plan.Initiators); i++ {
+			planInitiatorMap[plan.Initiators[i].PortName] = plan.Initiators[i]
+		}
+	}
+
+	// fetch the plan data to get password value.
+	// Passwords are not queryable so in order to maintain state it is taken from the plan.
+	stateInitiatorMap := make(map[types.String]models.InitiatorCreateModify)
+	if state != nil && len(state.Initiators) != 0 {
+		for i := 0; i < len(state.Initiators); i++ {
+			stateInitiatorMap[state.Initiators[i].PortName] = state.Initiators[i]
+		}
+	}
+
 	for _, initiator := range response.Initiators {
-		initiators = append(initiators, models.InitiatorCreateModify{
-			PortName: types.StringValue(initiator.PortName),
-			PortType: types.StringValue(string(initiator.PortType)),
-		})
+		initiatorModel := models.InitiatorCreateModify{}
+		if initiator.PortType != "iSCSI" || (initiator.ChapMutualUsername == "" && initiator.ChapSingleUsername == "") {
+			initiatorModel.PortName = types.StringValue(initiator.PortName)
+			initiatorModel.PortType = types.StringValue(string(initiator.PortType))
+		} else if initiator.ChapMutualUsername == "" {
+			initiatorModel.PortName = types.StringValue(initiator.PortName)
+			initiatorModel.PortType = types.StringValue(string(initiator.PortType))
+			if operation != operationRead {
+				initiatorModel.ChapSinglePassword = planInitiatorMap[types.StringValue(initiator.PortName)].ChapSinglePassword
+			} else {
+				initiatorModel.ChapSinglePassword = stateInitiatorMap[types.StringValue(initiator.PortName)].ChapSinglePassword
+			}
+			initiatorModel.ChapSingleUsername = types.StringValue(initiator.ChapSingleUsername)
+		} else {
+			initiatorModel.PortName = types.StringValue(initiator.PortName)
+			initiatorModel.PortType = types.StringValue(string(initiator.PortType))
+			if operation != operationRead {
+				initiatorModel.ChapSinglePassword = planInitiatorMap[types.StringValue(initiator.PortName)].ChapSinglePassword
+				initiatorModel.ChapMutualPassword = planInitiatorMap[types.StringValue(initiator.PortName)].ChapMutualPassword
+			} else {
+				initiatorModel.ChapSinglePassword = stateInitiatorMap[types.StringValue(initiator.PortName)].ChapSinglePassword
+				initiatorModel.ChapMutualPassword = stateInitiatorMap[types.StringValue(initiator.PortName)].ChapMutualPassword
+			}
+			initiatorModel.ChapMutualUsername = types.StringValue(initiator.ChapMutualUsername)
+			initiatorModel.ChapSingleUsername = types.StringValue(initiator.ChapSingleUsername)
+		}
+
+		initiators = append(initiators, initiatorModel)
 	}
 	state.Initiators = initiators
 	if operation == operationRead {
 		state.HostGroupID = types.StringValue(response.HostGroupID)
 	}
 }
+
+func (r resourceHost) planToServer(plan, state models.Host) *gopowerstore.HostModify {
```

Comment:
Can we have proper comments above the function implementation ?

Reply:
Added